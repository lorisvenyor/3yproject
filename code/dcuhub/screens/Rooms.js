'use strict'
import React, { Component, useEffect } from 'react';
import { Text, Image, TextInput, TouchableHighlight, StatusBar, FlatList, View, TouchableOpacity, Dimensions, LogBox, Alert } from 'react-native';
// import firebaseApp from './firebaseConfig.js';
import firebaseApp from '../config/firebasekeys';
import {SafeAreaView} from 'react-native-safe-area-context'
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view'
import styles from './ChatStyle';
import HeaderDisplay from '../shared/header'
import { Ionicons } from '@expo/vector-icons'; 


LogBox.ignoreAllLogs(['VirtualizedLists should never be nested inside plain ScrollViews with the same orientation'])

class Rooms extends Component {
    
    _isMounted = false

    static navigationOptions = {
        title: 'Rooms',
        headerShown: true
    }

    constructor(props) {
        super(props);
            var firebaseDB = firebaseApp.database();
            this.roomsRef = firebaseDB.ref('rooms'); // rooms collection
            this.state = {
            rooms: [],
            newRoom: ''
        }
    }

    componentDidMount() {
        this._isMounted = true
        
        if(this._isMounted){
            LogBox.ignoreLogs(['VirtualizedLists should never be nested']); // doing this to ignore the log box error for scroll view since scroll view is needed
            this.listenForRooms(this.roomsRef)
        }
    }
    
    componentWillUnmount(){
        this._isMounted = false
    }

    listenForRooms(roomsRef) {
        this._isMounted = true
        
        roomsRef.on('value', (dataSnapshot) => {
        var roomsFB = [];
        dataSnapshot.forEach((child) => {
            roomsFB.push({
            name: child.val().name,
            key: child.key
            });
        })
        if(this._isMounted){
            this.setState({ rooms: roomsFB })
        }
        // this.setState({ rooms: roomsFB })
        });
    }

    addRoom() {
        if (this.state.newRoom === '') {
            Alert.alert('Enter Chat Name', `Please enter a chat name.`)
        }else{
            this.roomsRef.push({ name: this.state.newRoom });
            this.setState({ newRoom: '' });
        }
    }

    openMessages(room) {
        this.props.navigation.navigate("Messages", {roomKey: room.key, roomName: room.name});
    }

    renderRow(item) {
        return (
        <TouchableOpacity style={{
            flex: 1,
            marginTop: 20,
            marginHorizontal: 30,
            padding: 15,
            borderRadius: 20,
            borderColor: 'orange',
            borderWidth: 2,
        }}
            onPress={() => this.openMessages(item)}
            
            onLongPress={() => 
                Alert.alert(
                    'Confirm',
                    `Delete chat room '${item['name']}'?`,
                    [
                      {text: 'Delete', onPress: () => this.roomsRef.child(item.key).remove()},
                      {
                        text: 'Cancel', onPress: () => console.log('OK Pressed'),
                      },
                    ],
                    {cancelable: false},
                  )
                }
            > 
            {/* here's where the chat name title go. so cut the length of the string if it goes beyond 30 */}
            {item.name.length < 30 ? <Text style={{color: 'orange', fontSize: 21}}>{item.name}</Text> :
            <Text style={{color: 'orange', fontSize: 20}}>{item.name.substring(0,30) + "..."}</Text>
            }
        </TouchableOpacity>
        )    
    }

  render() {
        return (
        <SafeAreaView style={styles.container}>
        <StatusBar style="auto"/>
        <KeyboardAwareScrollView>
        
        <HeaderDisplay title="MESSAGES"/>
            
            {/* the whole view */}
            <View style={{
                    flex: 1,
                    alignItems: 'center',
                    justifyContent: 'flex-start',
                    paddingBottom: 10,
                    marginTop: 10
            }}>        
                {/* enter a chat group name */}
                <View style={{
                        paddingTop: 8, 
                        fontSize: 25,
                        flexDirection: 'row',
                        }}>

                    <TextInput
                        style={{
                            flex: 1,
                            height: 40,
                            fontSize: 18,
                            color: 'black',
                            borderColor: '#25afcd',
                            borderWidth: 1.5,
                            borderRadius: 10,
                            paddingHorizontal: 20,
                            marginLeft: 20,
                            marginRight: 100,

                        }}
                        placeholder={"create a chat group"}
                        onChangeText={(text) => this.setState({newRoom: text})}
                        value={this.state.newRoom}
                    />
                    <TouchableHighlight
                        style={{
                            marginRight: 20
                        }}
                        onPress={() => this.addRoom()}
                        >
                        <Ionicons name="ios-create-outline" size={35} color="#25afcd" />
                    </TouchableHighlight>

                </View>

                <View style={{
                        flex: 1,
                        alignItems: 'flex-start',
                        justifyContent: 'center',
                        flexDirection: 'row',
                        marginTop: 30,
                        marginBottom: 20
                }}>
                <FlatList
                    data={this.state.rooms}
                    renderItem={({item}) => (this.renderRow(item)
                    )}/>
                </View>
    
            </View>

        </KeyboardAwareScrollView>
        </SafeAreaView>

        )
    }
}

export default Rooms

