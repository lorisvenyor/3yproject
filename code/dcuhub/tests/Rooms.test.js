import React from 'react'
import {create, act} from 'react-test-renderer'
import Rooms from '../screens/Rooms'


const screen = create(<Rooms/>)

test('Rooms Screen Snapshot', () => {
    expect(screen).toMatchSnapshot()
})
