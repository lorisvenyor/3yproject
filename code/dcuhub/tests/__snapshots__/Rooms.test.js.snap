// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rooms Screen Snapshot 1`] = `
<RNCSafeAreaView
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
      "height": 1334,
    }
  }
>
  <RCTScrollView
    automaticallyAdjustContentInsets={false}
    contentInset={
      Object {
        "bottom": 0,
      }
    }
    enableAutomaticScroll={true}
    enableOnAndroid={false}
    enableResetScrollToCoords={true}
    extraHeight={75}
    extraScrollHeight={0}
    getScrollResponder={[Function]}
    handleOnScroll={[Function]}
    keyboardDismissMode="interactive"
    keyboardOpeningTime={250}
    keyboardSpace={0}
    onScroll={[Function]}
    resetKeyboardSpace={[Function]}
    scrollEventThrottle={1}
    scrollForExtraHeightOnAndroid={[Function]}
    scrollIntoView={[Function]}
    scrollToEnd={[Function]}
    scrollToFocusedInput={[Function]}
    scrollToPosition={[Function]}
    showsVerticalScrollIndicator={true}
    update={[Function]}
    viewIsInsideTabBar={false}
  >
    <View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-around",
            "marginBottom": -10,
            "marginTop": -10,
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#25afcd",
              "borderColor": "black",
              "borderRadius": 20,
              "borderWidth": 0,
              "marginBottom": 30,
              "marginTop": 30,
              "paddingHorizontal": 20,
              "paddingVertical": 0,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 25,
                "padding": 8,
              }
            }
          >
            MESSAGES
          </Text>
        </View>
        <View>
          <Image
            resizeMode="center"
            source={1}
            style={
              Object {
                "height": 110,
                "width": 110,
              }
            }
          />
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "flex-start",
            "marginTop": 10,
            "paddingBottom": 10,
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "fontSize": 25,
              "paddingTop": 8,
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            placeholder="create a chat group"
            rejectResponderTermination={true}
            style={
              Object {
                "borderColor": "#25afcd",
                "borderRadius": 10,
                "borderWidth": 1.5,
                "color": "black",
                "flex": 1,
                "fontSize": 18,
                "height": 40,
                "marginLeft": 20,
                "marginRight": 100,
                "paddingHorizontal": 20,
              }
            }
            underlineColorAndroid="transparent"
            value=""
          />
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "marginRight": 20,
              }
            }
          >
            <Text />
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "flex-start",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "center",
              "marginBottom": 20,
              "marginTop": 30,
            }
          }
        >
          <RCTScrollView
            data={Array []}
            disableVirtualization={false}
            getItem={[Function]}
            getItemCount={[Function]}
            horizontal={false}
            initialNumToRender={10}
            keyExtractor={[Function]}
            maxToRenderPerBatch={10}
            onContentSizeChange={[Function]}
            onEndReachedThreshold={2}
            onLayout={[Function]}
            onMomentumScrollEnd={[Function]}
            onScroll={[Function]}
            onScrollBeginDrag={[Function]}
            onScrollEndDrag={[Function]}
            removeClippedSubviews={false}
            renderItem={[Function]}
            scrollEventThrottle={50}
            stickyHeaderIndices={Array []}
            updateCellsBatchingPeriod={50}
            viewabilityConfigCallbackPairs={Array []}
            windowSize={21}
          >
            <View />
          </RCTScrollView>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;
